File: day-1/historian-hysteria/java/historianHysteria.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.*;

public class historianHysteria {
    public static void main(String[] args) throws FileNotFoundException {
        System.out.println(partOne());
        System.out.println(partTwo());
    }

    private static int partOne() throws FileNotFoundException {
        File file = new File("input.txt");
        Scanner scanner = new Scanner(file);
        int total = 0;
        List<Integer> list1 = new ArrayList<>();
        List<Integer> list2 = new ArrayList<>();
        while(scanner.hasNextLine()) {
            String[] locationIds = scanner.nextLine().split(" {3}");
            list1.add(Integer.parseInt(locationIds[0]));
            list2.add(Integer.parseInt(locationIds[1]));
        }

        list1.sort(Integer::compareTo);
        list2.sort(Integer::compareTo);

        for (int i = 0; i < list1.size(); i++) {
            total += Math.abs(list1.get(i) - list2.get(i));
        }

        return total;
    }

    private static int partTwo() throws FileNotFoundException {
        File file = new File("input.txt");
        Scanner scanner = new Scanner(file);
        Map<Integer, Integer> rightListFrequencyMap = new HashMap<>(1000);
        List<Integer> leftList = new ArrayList<>();
        int similarityScore = 0;

        while (scanner.hasNextLine()) {
            String[] locationIds = scanner.nextLine().split(" {3}");
            int right = Integer.parseInt(locationIds[1]);
            rightListFrequencyMap.put(right, rightListFrequencyMap.getOrDefault(right, 0) + 1);
            leftList.add(Integer.parseInt(locationIds[0]));
        }

        for(int i : leftList) {
            similarityScore += i * rightListFrequencyMap.getOrDefault(i, 0);
        }

        return similarityScore;
    }
}

File: day-1/historian-hysteria/python/historianHysteria.py
def part_one():
    lines = []

    with open("input.txt") as file:
        lines = file.readlines()

    first_list = []
    second_list = []

    for i in range(len(lines)):
        first_list.append(int(lines[i].split()[0]))
        second_list.append(int(lines[i].split()[1]))

    first_list.sort()
    second_list.sort()

    total = 0
    for i in range(len(first_list)):
        total += abs(first_list[i] - second_list[i])

    return total

def part_two():
    with open("input.txt") as file:
        lines = file.readlines()

    right_frequency_map = {}
    left_list = []
    similarityScore = 0

    for i in range(len(lines)):
        left_list.append(int(lines[i].split()[0]))
        right = int(lines[i].split()[1])
        right_frequency_map.update({right : right_frequency_map.get(right, 0) + 1})

    for i in left_list:
        similarityScore += i * right_frequency_map.get(i, 0)

    return similarityScore


print(part_one())
print(part_two())